### PR TITLE
crossfilter.remove() to take a test function argument 

### PR DESCRIPTION
**Pull request by tehsenaus**
This change addresses https://github.com/crossfilter/crossfilter/issues/6

### CODE REVIEW

**Comment by esjewett** [comment]
Excellent - thanks! I'm like a month behind on pull requests (sorry @alexcampana). I will try to get to these in the next week.

**Comment by tehsenaus** [comment]
Great, thanks. Let me know if you need any help with those other PRs, I feel I've built up some level of understanding of the library's internals through working on this.

**Comment by esjewett** [comment]
If you want to read through pull requests and tests, and run the tests locally to validate that they are behaving correctly, that would certainly be helpful. I'm going to try to tackle these today, so doing it now might be a bit duplicative.

**Comment by esjewett** [comment]
@tehsenaus Thanks again for this valuable contribution. I've reviewed this and I'm going to mull it over for another day. It looks good but the internals of Crossfilter are very difficult to reason about. I think I would like to see the following changes/additions to your tests:

- Un-filter in the middle of your filter test (when you still have 2 records) and make sure that group totals are still correct (this is where my attempted change failed - it removed both records matching the predicate and records outside of the current filter - so I'm inclined to want to test for that condition)
- A test based on array dimensions

But overall I think this looks very good and I'm not too concerned that those tests will fail.

Would you be up for adding those test conditions? If not, I will try to do it tomorrow. Could you also remove the `crossfilter.js` and `crossfilter.min.js` files from your pull request? That will get rid of the branch conflicts. I'll handle generating the those build artifacts and pushing a release once we get this sorted out.

**Comment by tehsenaus** [comment]
Yep this sounds good - I'll do as you suggested. I'll try to take a look at this today.

**Comment by tehsenaus** [comment]
I'm looking at this now - I think the "can remove records while filtering" test covers your first case, though I've added a few assertions there.

There's a bug with the array dimensions though. Will this be fixed by https://github.com/crossfilter/crossfilter/pull/83? Which PR is going to land first?

**Comment by esjewett** [comment]
I guess it does, doesn't it? You're right.

I'm not sure which should land first. I'm thinking probably #83 should. Would you be OK with me landing that one on master then re-checking this one?

**Comment by esjewett** [comment]
Ok, I'll do that shortly. They are not so bad conceptually. It's just putting each record into more than one group if necessary and maintaining a dimension index that can be longer (because each record can appear in more than one place in the index), IIRC. But because of the current internal architecture that causes branching code paths all over the place and it is a bit of a mess. There was a refactoring that someone was working on at one point but it never landed. It's still on my to-do list along with a rework of the API to be asynchronous for a 2.0 release, but you know how that goes...

**Comment by esjewett** [comment]
I've now merged #83 and this has predictably caused a couple of merge conflicts. Will you be able to update against the latest master @tehsenaus? I'll try not to step on you again here. Thanks again for working on this.

**Comment by tehsenaus** [comment]
Done!

I agree about the current code structure, the large files are quite hard to follow. It'd be great to refactor this to modern JS standards, I think it could be done without any breaking changes to the API.

Have you considered moving to TypeScript? Could work well for a low-level library like this.

**Comment by esjewett** [comment]
Thank you! Yes, I would love to move to either TypeScript (which I have some experience in) or one of the pure functional languages like Elm or Purescript (which I would have to learn from scratch). However, I just haven't had time for the full rewrite that would entail. Would you be interesting in collaborating on that?

**Comment by tehsenaus** [comment]
Thanks for the merge!

Yes I'd be happy to collaborate on that. A TypeScript port shouldn't take too much rewriting, just refactoring really.
